exonum-java-binding/testkit/src/test/java/com/exonum/binding/testkit/TestKitTest.java

```diff
@@ -31,115 +33,222 @@
+        ARTIFACT_ID.toString());
   }
 
   @Test
-  void createTestKitWithBuilderForMultipleDifferentServices() {
+  void createTestKitWithDifferentDeployedArtifactAndCreatedServiceThrows() {
+    Class<IllegalArgumentException> exceptionType = IllegalArgumentException.class;
+    TestKit.Builder testKitBuilder = TestKit.builder()
+        .withDeployedArtifact(ARTIFACT_ID, ARTIFACT_FILENAME)
+        .withArtifactsDirectory(artifactsDirectory);
+    IllegalArgumentException thrownException = assertThrows(exceptionType,
+        () -> testKitBuilder.withService(ARTIFACT_ID_2, SERVICE_NAME_2, SERVICE_ID_2));
+    assertThat(thrownException.getMessage())
+        .isEqualTo("Service %s should be deployed first in order to be created",
+        ARTIFACT_ID_2.toString());
+  }
+
+  @Test
+  void createTestKitWithoutArtifactsDirectoryThrows() {
+    Class<NullPointerException> exceptionType = NullPointerException.class;
+    TestKit.Builder testKitBuilder = TestKit.builder()
+        .withDeployedArtifact(ARTIFACT_ID, ARTIFACT_FILENAME)
+        .withService(ARTIFACT_ID, SERVICE_NAME, SERVICE_ID);
+    NullPointerException thrownException = assertThrows(exceptionType, testKitBuilder::build);
+    assertThat(thrownException.getMessage()).isEqualTo("Artifacts directory was not set.");
+  }
+
+  @Test
+  void createTestKitWithBuilderForSingleServiceWithCustomConfiguration() {
     try (TestKit testKit = TestKit.builder()
-        .withService(TestServiceModule.class)
-        .withService(TestServiceModule2.class)
+        .withDeployedArtifact(ARTIFACT_ID, ARTIFACT_FILENAME)
+        .withService(ARTIFACT_ID, SERVICE_NAME, SERVICE_ID, DEFAULT_CONFIGURATION)
+        .withArtifactsDirectory(artifactsDirectory)
         .build()) {
-      TestService service = testKit.getService(TestService.SERVICE_ID, TestService.class);
-      checkTestServiceInitialization(testKit, service);
-      TestService2 service2 = testKit.getService(TestService2.SERVICE_ID, TestService2.class);
-      checkTestService2Initialization(testKit, service2);
+      checkTestServiceInitialization(testKit, SERVICE_NAME, SERVICE_ID);
     }
   }
 
+  @ParameterizedTest
+  @ValueSource(ints = {-1, MAX_SERVICE_INSTANCE_ID + 1})
+  void createTestKitWithInvalidServiceId(int invalidServiceId) {
+    Class<IllegalArgumentException> exceptionType = IllegalArgumentException.class;
+    TestKit.Builder testKitBuilder = TestKit.builder()
+        .withDeployedArtifact(ARTIFACT_ID, ARTIFACT_FILENAME)
+        .withArtifactsDirectory(artifactsDirectory);
+    assertThrows(exceptionType,
+        () -> testKitBuilder.withService(ARTIFACT_ID, SERVICE_NAME, invalidServiceId));
+  }
+
   @Test
-  void createTestKitWithBuilderForMultipleDifferentServicesVarargs() {
+  void createTestKitWithBuilderForMultipleDifferentServices() throws Exception {
+    // Create artifact for TestService2
+    createArtifact(artifactsDirectory.resolve(ARTIFACT_FILENAME_2));
```

Comment:
It is broken: this method writes ARTIFACT_ID, but then specifies ARTIFACT_ID_2:
     .withDeployedArtifact(ARTIFACT_ID_2, ARTIFACT_FILENAME_2)

runtime will reject that